src/frontend/serialization.rs

```diff
@@ -1,109 +1,276 @@
-    backends::plonky2::mock::{mainpod::MockMainPod, signedpod::MockSignedPod},
-    frontend::{Error, MainPod, SignedPod, Statement},
-    middleware::{containers::Dictionary, Key, PodId, Value},
+    backends::plonky2::{
+        basetypes::{C, D},
+        circuits::mainpod::MainPodVerifyCircuit,
+        mainpod::{
+            pad_statement, MainPod as Plonky2MainPod, MainPodProof, Statement as BackendStatement,
+        },
+        mock::{mainpod::MockMainPod, signedpod::MockSignedPod},
+        primitives::signature::{Signature, VP},
+        signedpod::SignedPod as Plonky2SignedPod,
+    },
+    frontend::{MainPod, SignedPod},
+    middleware::{
+        self, containers::Dictionary, serialization::ordered_map, AnchoredKey, Key, Params, PodId,
+        Statement, StatementArg, Value, F, SELF,
+    },
 };
 
+#[derive(Serialize, Deserialize, JsonSchema)]
+pub enum SignedPodType {
+    Signed,
+    MockSigned,
+}
+
 #[derive(Serialize, Deserialize, JsonSchema)]
 #[serde(rename_all = "camelCase")]
-#[schemars(title = "SignedPod")]
-pub struct SignedPodHelper {
+#[schemars(rename = "SignedPod")]
+pub struct SerializedSignedPod {
+    id: PodId,
+    #[serde(serialize_with = "ordered_map")]
     entries: HashMap<Key, Value>,
     proof: String,
-    pod_class: String,
-    pod_type: String,
+    pod_type: SignedPodType,
 }
 
-impl TryFrom<SignedPodHelper> for SignedPod {
-    type Error = Error;
-
-    fn try_from(helper: SignedPodHelper) -> Result<SignedPod, Self::Error> {
-        if helper.pod_class != "Signed" {
-            return Err(Error::custom("pod_class is not Signed"));
-        }
-        if helper.pod_type != "Mock" {
-            return Err(Error::custom("pod_type is not Mock"));
-        }
-
-        let dict = Dictionary::new(helper.entries.clone())?.clone();
-        let pod = MockSignedPod::new(PodId(dict.commitment()), helper.proof, dict.kvs().clone());
+#[derive(Serialize, Deserialize, JsonSchema)]
+pub enum MainPodType {
+    Main,
+    MockMain,
+}
 
-        Ok(SignedPod {
-            pod: Box::new(pod),
-            kvs: helper.entries,
-        })
-    }
+#[derive(Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+#[schemars(rename = "MainPod")]
+pub struct SerializedMainPod {
+    id: PodId,
+    public_statements: Vec<Statement>,
+    proof: String,
+    params: Params,
+    pod_type: MainPodType,
 }
 
-impl From<SignedPod> for SignedPodHelper {
+impl From<SignedPod> for SerializedSignedPod {
     fn from(pod: SignedPod) -> Self {
-        SignedPodHelper {
+        SerializedSignedPod {
+            id: pod.id(),
             entries: pod.kvs,
             proof: pod.pod.serialized_proof(),
-            pod_class: "Signed".to_string(),
-            pod_type: "Mock".to_string(),
+            pod_type: if (&*pod.pod as &dyn Any)
+                .downcast_ref::<Plonky2SignedPod>()
+                .is_some()
+            {
+                SignedPodType::Signed
+            } else {
```

Comment:
I would suggest doing a downcast to `MockSignedPod` here in an `else if`, and have a final `else` with an `unreachable!()`, so that this code is robust against future new signed pod types.  Otherwise the MockSigned would be the fallback.
The same applies to `impl From<MainPod> for SerializedMainPod`

Reply:
Done